### PR TITLE
resize-limiter: add additional check for low-hz monitors

### LINKED ISSUE #4542: 9f20a15 Introduces subtle stuttering when moving floating windows
### Hyprland Version

<details> <summary>System/Version info</summary>

```sh
Hyprland, built from branch main at commit 8b1069b33051dbc3a133444dc6ce7385cdc51377 dirty (flake.lock: update).
Date: Sat Jan 27 09:31:39 2024
Tag: v0.34.0-68-g8b1069b3

flags: (if any)


System Information:
System name: Linux
Node name: zombie
Release: 6.7.1-arch1-1
Version: #1 SMP PREEMPT_DYNAMIC Sun, 21 Jan 2024 22:14:10 +0000


GPU information:
0000:00:02.0 VGA compatible controller [0300]: Intel Corporation TigerLake-LP GT2 [Iris Xe Graphics] [8086:9a49] (rev 01) (prog-if 00 [VGA controller])


os-release: NAME="Arch Linux"
PRETTY_NAME="Arch Linux"
ID=arch
BUILD_ID=rolling
ANSI_COLOR="38;2;23;147;209"
HOME_URL="https://archlinux.org/"
DOCUMENTATION_URL="https://wiki.archlinux.org/"
SUPPORT_URL="https://bbs.archlinux.org/"
BUG_REPORT_URL="https://gitlab.archlinux.org/groups/archlinux/-/issues"
PRIVACY_POLICY_URL="https://terms.archlinux.org/docs/privacy-policy/"
LOGO=archlinux-logo


plugins:
```

</details>

### Bug or Regression?

Regression

### Description

9f20a15955697b0f52740c1e0180acdafa071f79 introduces subtle stutter/jitter(?) when moving a floating window.

Reverting that commit fixes the issue.

### How to reproduce

\> build latest
\> make a window float
\> move it around
\> stutters subtly
\> die of embarrassment

### Crash reports, logs, images, videos

_No response_

### LINKED ISSUE #4669: Sluggish moving windows in 0.35
### Hyprland Version

<details> <summary>System/Version info</summary>

```sh
Hyprland, built from branch HEAD at commit 84ab8d11e8951a6551d1e1bf87796a8589da6d47  (props: bump ver to 0.35.0).
Date: Mon Feb 5 01:59:02 2024
Tag: v0.35.0
```

</details>

### Bug or Regression?

Bug

### Description

The video faintly shows this effect, but I compiled 0.34 and confirmed that the bug is there. The bug consists in statting when dragging a window. I noticed that the more "fat" the application is, the more lags occur. Strangely enough, changing desktops  is smooth.

### How to reproduce

Drag and drop windows

### Crash reports, logs, images, videos

https://youtu.be/iuprk8g_hDc

### PR DESCRIPTION
**Pull request by bvr-yr**
#### Describe your PR, what does it fix/add?

adding 5-6 ms of delay can fix render lag on mouse movement events in some weird combo `mouse + 60hz display`

should fix #4542 
should fix #4669

#### Is there anything you want to mention? (unchecked code, possible bugs, found problems, breaking compatibility, etc.)

nope

#### Is it ready for merging, or does it need work?

ready

### CODE REVIEW

**Comment by vaxerski** [comment]
unsure how this is supposed to fix anything tbh

**Comment by bvr-yr** [comment]
try forcing 60hz refresh rate and move wayland window with `bindm`, it moves like at 30fps
maybe it's because old hardware but i doubt, cause issue OP have relativly new (tigerlake) intel gpu

this just returns things like they were before https://github.com/hyprwm/Hyprland/commit/9f20a15955697b0f52740c1e0180acdafa071f79 not breaking already done fixes

if there is different solution, sure, but atleast i tried :D

**Comment by vaxerski** [comment]
christ 60hz is disgusting.

Anyways tried and can't see any issues.

**Comment by bvr-yr** [comment]
what can i do it's builtin laptop display :p
with plugged 120hz moni its okay rendering on both

well, it might be intel issue, as i'm not only one having this on intel
seen some reports here considering slow gpu wakeups even with forced maxed clock rate. without that forcing animations are noway

**Comment by bvr-yr** [comment]
@vaxerski i think i found the root cause of the problem
may i ask what mouse are u using?

**Comment by burein-ita** [comment]
Just tested, don't know why but this indeed fixes #4542

**Comment by bvr-yr** [comment]
i suppose that's because of there are too much returning instead of render in 16.66666ms window (60hz) for default, not hi-res input devices, and since that might be not the case for the majority of userbase, issue is pretty niche

@vaxerski is adding cfg option viable? or i could think on some another logic in checker

**Comment by vaxerski** [comment]
hm. I am still unsure what the best course of action is here. Maybe update the cursor pos all the time, and update resize status in a preRender hook?

**Comment by bvr-yr** [comment]
there's another related issue #4669
i doubt i can do more than adding cfg option, cause it will be a guessing from my side

feel free to close if you are planning to fix it in natural way, i can continue patching if needed

**Comment by vaxerski** [comment]
I have no idea why that happens, but making the resize _inherently more choppy_ is not the way to do it.

**Comment by bvr-yr** [comment]
could u please run this snippet from shell
```bash
declare -i n=0 total=0
unset time0 time1 m avg
(( ZSH_VERSION )) && m=1.0 || m=1
wev -f wl_pointer:motion | while read -r line; do
	n+=1
	time0=${line%;*}
	time0=${time0#*time: }
	(( delta = time0 - ${time1:-$time0} ))
	(( total += delta < 16 ? delta : avg ))
	(( avg = total * m / n ))
	echo "time:$time0  ms:$delta  avg:$avg"
	time1=${line%;*}
	time1=${time1#*time: }
done
```

move mouse a bit and paste few last lines

**Comment by vaxerski** [comment]
finna get a few trojans

**Comment by bvr-yr** [comment]
nice, u also have shitty mouse

look, can we do like this?
as been reported in https://github.com/hyprwm/Hyprland/issues/4542 and https://github.com/hyprwm/Hyprland/issues/4669 issue arises mostly on 60hz
now, if 60hz, there is additional check making it not return 1 extra time and draw instead when input events are too slow

if mouse sends ~1ms (like g603 which i tested), maybe it will skip 1 or 2 returns out of ~15

although, 75hz been mentioned once, i can't see it stuttering, atleast on headless
but if needed, we can just lower bound to `13.0`

**Comment by vaxerski** [comment]
I can't see any stutters even on 1khz

**Comment by bvr-yr** [comment]
but u won't, it's only 60hz and 7-8ms

**Comment by bvr-yr** [comment]
that's the problem, when both conditions are met:
```
return
return
draw
return
return
draw
```

now its
```
return
draw
return
draw
```

File: src/layout/IHyprLayout.cpp [hunk 1]
```diff
@@ -313,9 +316,21 @@ void IHyprLayout::onMouseMove(const Vector2D& mousePos) {
...
+    if (m_iMouseMoveEventCount == 1)
+        totalMs = 0;
+
+    totalMs += MSDELTA < MSMONITOR ? MSDELTA : std::round(totalMs * 1.0 / m_iMouseMoveEventCount);
+    m_iMouseMoveEventCount += 1;
+
     if ((abs(TICKDELTA.x) < 1.f && abs(TICKDELTA.y) < 1.f) ||
-        (std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::high_resolution_clock::now() - TIMER).count() <
-         1000.0 / g_pHyprRenderer->m_pMostHzMonitor->refreshRate))
+        (TIMERDELTA < MSMONITOR && (MSMONITOR < 16.0 || std::clamp(MSMONITOR - TIMERDELTA, 0.0, MSMONITOR) > totalMs * 1.0 / m_iMouseMoveEventCount)))
```
Comment (vaxerski) [inline src/layout/IHyprLayout.cpp:333]: this is very confusing, some boolean variables or comments would help.

Comment (bvr-yr) [reply src/layout/IHyprLayout.cpp:333]: smth like
```cpp
// if we face 60hz monitor, let's check wether remaing time before refresh is more than average input-events sending rate
```
ok?

Comment (vaxerski) [reply src/layout/IHyprLayout.cpp:333]: see, your code is so confusing you even confused yourself. I still recommend booleans.

`MSMONITOR > 16.0` if you wanna find 60Hz monitors. <16.0 targets everything _but_ 60hz panels.

Comment (bvr-yr) [reply src/layout/IHyprLayout.cpp:333]: wait, we wan't orinial limiter to work, then it checks if monitor is **not** 60hz and returns, skipping average timing check
if it is 60hz (`MSMONITOR < 16.0` fails), perform one more check

Comment (vaxerski) [reply src/layout/IHyprLayout.cpp:333]: see, too confusing.

**Review by vaxerski — APPROVED** [review]
united we stand

**Comment by Sinaaaa** [comment]
I thought the bad floating window movement performance was on my haswell igpu, but hey this thing seems to have made quite a positive difference.

**Comment by bvr-yr** [comment]
yeah, it was confusing. initially i thought the same (lulswell too), but people on Tigerlake had same problem, and it is pure luck i have 2 monitors and switchable mouse to catch the problem

glad to help
no more `yay --editmenu` thank god

**Comment by kemzops** [comment]
Thanks you !! i was trying to fix this issue but i failed, and this actually work.

**Pull request merged**